### DISCUSSION TITLE
How to use XSPEC models as a table model for PHA analysis

### DISCUSSION THREAD

**Discussion by DougBurke** [sherpa/sherpa#1849] [Show and tell]
This is based on a question from a local user, who basically had the following very-simplified code to show problems with using an XSPEC model as a table model

```
from matplotlib import pyplot as plt
from sherpa.astro import ui

ui.load_pha('spec.pi')
rmf = ui.get_rmf(); arf = ui.get_arf()
pl= ui.powerlaw.pl; pl.ampl=1.2e-13; pl.index=-3.
ui.set_source(pl)

## check plot
ui.plot_source(); plt.yscale('log'); plt.show() ## tops out ~1e1

## write out the data
tmp= ui.get_source_plot(); xlo=tmp.xlo; xhi=tmp.xhi; y=tmp.y
ui.save_arrays('tmp.txt', [(xlo+xhi)/2,y], ['keV','photon/sec/cm^2/keV'], ascii=True, clobber=True)

## load the table model

ui.load_table_model('arr','tmp.txt')
ui.set_source(arr)
ui.plot_fit(); plt.yscale('log'); plt.show() ## model = 100x too high
```

The problem is that the table model values are too high - here's the `plot_fit` call for the the powerlaw model and then for the table model, which we would expect to be essentially the same:

![orig](https://github.com/sherpa/sherpa/assets/224638/3aac2e95-6f9b-4978-96e2-311d662db91d)
![wrong](https://github.com/sherpa/sherpa/assets/224638/4388e4fa-fa52-4a17-b052-9c108ce633b9)

We have several issues here:
- the table model doesn't handle integrated models, which explains the 100 times factor (as this response has bins of with 0.01 keV)
- the table model will interpolate things, which seems to come into play here.

We can improve things, but they may cause some down-stream problems:

1. rather than write out the per-keV data, write out "integrated" data

```
tmp= ui.get_source_plot(); xlo=tmp.xlo; xhi=tmp.xhi; y=tmp.y
ui.save_arrays('tmp.txt', [(xlo+xhi)/2, y * (xho - xlo)], ['emid','source'], ascii=True, clobber=True)
```

2. only read in the model values, to trigger a different behavior in the table model (to avoid the interpolation change)

```
ui.load_table_model('tmod', 'tmp.txt', ncols=1, colkeys=["SOURCE"])
```

Note here setting `ncols` to 1 and to say what column you want (there's a possibility that `colkeys` is not needed but I don't want to guarantee that as the code pathway I'd have to check to ensure this is too long)

This should mean that the model will now be handled correctly. However, there could well be trouble if the data is grouped (but that would be a problem even without my suggested change).

An alternative is to write the data out as an XSPEC table model, but we currently have no way to do this in Sherpa (or CIAO).

I'm not sure what would be needed to handle this sensibly automatically (other than to go via an XSPEC table model). Needs thinking about.

**Comment by DougBurke** [comment]
I should add, this is only for "additive" style models - such as the power-law model used here. If a purely multiplicative model were written ou then the first step would write out `y` and not `y * bin-width`.

**Comment by anetasie** [comment]
@DougBurke This is an interesting case and yes, we do not have a handle on the models with two columns and integrated values over the bin and `load_xstable_model' requires the FITS file, so a model defines as an array cannot be input.

`load_table-model' supports two 'dstype' data Data1D and Data1DInt and for the PHA data sets with RMF/ARF the table model selects the integrated data set as default, so the model values 'y*bin-width' are cacluated. There is no information about the model being already integrated on the same grid as the PHA/ARF/RMF. This might be another regrid case.

**Comment by DougBurke** [comment]
I will note that in the Sherpa 4.16 release we added the `sherpa.astro.io.xstable` module - https://sherpa.readthedocs.io/en/4.16.0/overview/astro_io_xstable.html - which can be used to write out XSPEC table models.
